providers/etcd/shedlock-provider-etcd/src/main/java/net/javacrumbs/shedlock/provider/etcd/EtcdLockProvider.java

```diff
@@ -0,0 +1,198 @@
+import net.javacrumbs.shedlock.core.ClockProvider;
+import net.javacrumbs.shedlock.core.LockConfiguration;
+import net.javacrumbs.shedlock.core.LockProvider;
+import net.javacrumbs.shedlock.core.SimpleLock;
+import net.javacrumbs.shedlock.support.LockException;
+import net.javacrumbs.shedlock.support.annotation.NonNull;
+
+import java.math.BigDecimal;
+import java.math.RoundingMode;
+import java.time.Duration;
+import java.time.Instant;
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+
+import static net.javacrumbs.shedlock.support.Utils.getHostname;
+import static net.javacrumbs.shedlock.support.Utils.toIsoString;
+
+public class EtcdLockProvider implements LockProvider {
+
+    private static final String KEY_PREFIX = "job-lock";
+    // prepared multi-environment support
+    static final String ENV_DEFAULT = "default";
+
+    private final EtcdTemplate etcdTemplate;
+
+    public EtcdLockProvider(@NonNull Client client) {
+        this.etcdTemplate = new EtcdTemplate(client);
+    }
+
+    @Override
+    @NonNull
+    public Optional<SimpleLock> lock(@NonNull LockConfiguration lockConfiguration) {
+        String key = buildKey(lockConfiguration.getName(), ENV_DEFAULT);
+        return new EtcdLock(key, etcdTemplate, lockConfiguration).lease();
+    }
+
+    private static final class EtcdLock extends AbstractSimpleLock {
+        public static final BigDecimal MILLIS_IN_SECOND = BigDecimal.valueOf(1000);
+        private final String key;
+        private Long successLeaseId;
+        private final EtcdTemplate etcdTemplate;
+
+        private EtcdLock(String key, EtcdTemplate etcdTemplate, LockConfiguration lockConfiguration) {
+            super(lockConfiguration);
+            this.key = key;
+            this.etcdTemplate = etcdTemplate;
+        }
+
+        EtcdLock successLeaseId(Long successLeaseId) {
+            this.successLeaseId = successLeaseId;
+            return this;
+        }
+
+        Optional<SimpleLock> lease() {
+            Long leaseId = etcdTemplate.createLease(getSecondsUntil(lockConfiguration.getLockAtMostUntil()));
+
+            Optional<Long> lockSuccess = etcdTemplate.tryToLock(key, buildValue(), leaseId);
+            if (lockSuccess.isPresent()) {
+                return lockSuccess.map(this::successLeaseId);
+            } else {
+                etcdTemplate.revoke(leaseId);
+                return Optional.empty();
+            }
+        }
+
+        @Override
+        public void doUnlock() {
+            long keepLockFor = getSecondsUntil(lockConfiguration.getLockAtLeastUntil());
+
+            // lock at least until is in the past
+            if (keepLockFor <= 0) {
+                try {
+                    etcdTemplate.revoke(successLeaseId);
+                } catch (Exception e) {
+                    throw new LockException("Can not remove node", e);
+                }
+            } else {
+                Long leaseId = etcdTemplate.createLease(keepLockFor);
```

Comment:
What about the old lease?

Reply:
Should be timed out and then removed, eventually.